sootup.callgraph/src/main/java/sootup/callgraph/AbstractCallGraphAlgorithm.java

```diff
@@ -556,4 +543,87 @@ public static Optional<SootMethod> findConcreteMethod(
             + " and in its superclasses and interfaces");
     return Optional.empty();
   }
+
+  protected static Optional<SootMethod> findMethodInHierarchy(
+      @NonNull View view,
+      @NonNull SootClass sootClass,
+      @NonNull MethodSubSignature targetMethodSignature) {
+    SootMethod target = sootClass.getMethod(targetMethodSignature).orElse(null);
+    // check current class
+    if (target != null) {
+      return Optional.of(target);
+    }
+
+    ClassType superClassType = sootClass.getSuperclass().orElse(null);
+    // does not have a superclass
+    if (superClassType == null) {
+      return Optional.empty();
+    }
+    SootClass superclass = view.getClass(superClassType).orElse(null);
+    // superclass is mot in the view
+    if (superclass == null) {
+      return Optional.empty();
+    }
+
+    // method isn't found, continue with the superclass
+    return findMethodInHierarchy(view, superclass, targetMethodSignature);
+  }
+
+  /**
+   * Searches the default method that would be used as a target of the given SootClass and
+   * MethodSubSignature. All interfaces are checked and it returns that SootMethod object of the
+   * interface which is the subtype of all possible fitting default method.
+   *
+   * @param view The view contains the hierarchy information and the classes and methods
+   * @param sootClass the sootClass which defines the start of the search
+   * @param defaultSignature the method-subsignature which defines the target method
+   * @return An Optional containing the default method or an empty Optional if there is no default
+   *     method, or a superclass is not in the view.
+   */
+  protected static Optional<SootMethod> findDefaultMethod(
+      @NonNull View view,
+      @NonNull SootClass sootClass,
+      @NonNull MethodSubSignature defaultSignature) {
+
+    // get all possible default method targets
+    List<? extends SootMethod> fittingDefaultMethods =
+        view.getTypeHierarchy()
+            .implementedInterfacesOf(sootClass.getType())
+            .flatMap(
+                classType ->
+                    view
+                        .getMethod(
+                            view.getIdentifierFactory()
+                                .getMethodSignature(classType, defaultSignature))
+                        .stream())
+            .toList();
+
+    SootMethod defaultMethod = null;
+    // find default method
+    if (!fittingDefaultMethods.isEmpty()) {
+      for (SootMethod fittingDefaultMethod : fittingDefaultMethods) {
```

Comment:
instead of having an additional loop after collecting everything into a list: lets put that into a [reduce](https://www.baeldung.com/java-stream-reduce) call?